pil-analyzer/src/type_unifier.rs

```diff
@@ -127,22 +110,32 @@ impl Unifier {
             .insert(bound);
     }
 
-    fn add_substitution(&mut self, type_var: String, ty: Type) -> Result<(), String> {
+    fn add_substitution(&mut self, type_var: String, mut ty: Type) -> Result<(), String> {
+        self.substitute(&mut ty);
+        //        println!("Adding substitution: {type_var} -> {ty}");
         if ty.contains_type_var(&type_var) {
             return Err(format!(
                 "Cannot unify types {ty} and {type_var}: They depend on each other"
             ));
         }
-        let subs = [(type_var.clone(), ty.clone())].into();
 
         for bound in self.type_var_bounds(&type_var) {
             self.ensure_bound(&ty, bound)?;
         }
 
-        self.substitutions
```

Comment:
Who could have thought that it is bad for performance to iterate over all substitutions whenever a new one is added?